src/fhirtypes/common.ts

```diff
@@ -58,6 +58,91 @@ export function setPropertyOnDefinitionInstance(
   setPropertyOnInstance(instance, pathParts, assignedValue, fisher);
 }
 
+export function createUsefulSlices(
+  instanceDef: StructureDefinition | ElementDefinition | InstanceDefinition,
+  instanceOfStructureDefinition: StructureDefinition,
+  ruleMap: Map<string, { pathParts: PathPart[] }>,
+  fisher: Fishable
+) {
+  ruleMap.forEach(({ pathParts }, path) => {
+    const nonNumericPath = path.replace(/\[[-+]?\d+\]/g, '');
+    const element = instanceOfStructureDefinition.findElementByPath(nonNumericPath, fisher);
+    if (element) {
+      // go through the parts, and make sure that we have a useful index, and maybe a named slice
+      let current: any = instanceDef;
+      let currentPath = '';
+      for (const [i, pathPart] of pathParts.entries()) {
+        currentPath += `${currentPath ? '.' : ''}${pathPart.base}`;
+        const currentElement = instanceOfStructureDefinition.findElementByPath(currentPath, fisher);
+        const key =
+          pathPart.primitive && i < pathParts.length - 1 ? `_${pathPart.base}` : pathPart.base;
+
+        let index = getArrayIndex(pathPart);
+        let sliceName: string;
+        if (index == null) {
+          // if we are on an array element, treat no index as a 0 index
+          const baseIsArray =
+            currentElement?.base?.max != null &&
+            currentElement.base.max !== '0' &&
+            currentElement.base.max !== '1';
+          if (baseIsArray) {
+            index = 0;
+          }
+        }
+        if (index != null) {
+          // If the array doesn't exist, create it
+          if (current[key] == null) {
+            current[key] = [];
+          }
+          sliceName = pathPart.brackets ? getSliceName(pathPart) : null;
+          if (sliceName) {
+            const sliceIndices: number[] = [];
+            // Find the indices where slices are placed
+            const sliceExtensionUrl = fisher.fishForMetadata(sliceName)?.url;
```

Comment:
What happens if the slice is not on `extension`, but `sliceName` just happens to match an extension id?  Would that cause an issue here? or are we OK because we only use this to look for a match on url, so in that case there would not be a match on URL anyway?

I think we're probably OK. But I wanted to think this through out loud (virtually).